### ISSUE TITLE
[BUG] Upgrade fails because upgrade-repo VM is not becoming healthy and gets deleted again

### ISSUE THREAD

**Issue by bashofmann** [harvester/harvester#2782]
**Describe the bug**
The Harvester cluster was installed with 1.0.2 and is imported into a Rancher 2.6.8. Now it should be updated to 1.0.3, but the update fails when starting the upgrad-repo VM.

**To Reproduce**
Start an update from 1.0.2 to 1.0.3. The update starts and the upgrade download finishes successfully after a few minutes. The upgrade-repo VM is being started and also comes up. In Harvester you can see that the VM is running, but not healthy. In Longhorn you can see that a VM disk volume was successfully created and mounted.
You can see the following events on the VM:

```
Readiness probe failed: Get "http://10.52.3.3:80/harvester-iso/harvester-release.yaml": dial tcp 10.52.3.3:80: connect: no route to host
Readiness probe failed: Get "http://10.52.3.3:80/harvester-iso/harvester-release.yaml": dial tcp 10.52.3.3:80: connect: connection refused
```

`10.52.3.3` is the IP of the upgrade-repo VM.

![Bildschirmfoto 2022-09-19 um 10 33 09](https://user-images.githubusercontent.com/243056/191052064-d2dc7841-5150-4ce7-81f2-37c2ebd53c3b.png)

Then the VM is deleted again. The upgrade fails with a `502 Host not found` error.

![Bildschirmfoto 2022-09-19 um 10 33 27](https://user-images.githubusercontent.com/243056/191052003-286b2534-861c-47f6-88af-c9c3332f9964.png)

**Expected behavior**
The update successfully runs through.

**Support bundle**
Can be provided internally.

**Environment**
 - Harvester ISO version:  1.0.2
 - Underlying Infrastructure: 
 4 nodes, each 64 cores, 756gb ram. 2TB disks. Usage of all resources very low.
 Internet access through HTTP proxy. HTTP proxy is configured within Harvester and working. The HTTP proxy does not intercept or break TLS. no_proxy is configured will all internal ip ranges and TLDs. We also added `longhorn-system` and `cattle-system` manually to rule out https://github.com/harvester/harvester/issues/2436.

**Comment by bashofmann** [comment]
According to the logs in the support bundle, there are `apply-system-agent-upgrader` jobs failing with this:

```
2022-09-19T08:36:16.202270909Z + eval 'v="$HTTP_PROXY"'
2022-09-19T08:36:16.202289376Z + v=http://10.x.x.x:8080/
2022-09-19T08:36:16.202297040Z + '[' -z http://10.x.x.x:8080/ ]
2022-09-19T08:36:16.202302932Z + var=HTTPS_PROXY
2022-09-19T08:36:16.202309944Z + val=
2022-09-19T08:36:16.202324403Z + eval 'v="$HTTPS_PROXY"'
2022-09-19T08:36:16.202334984Z + v=http://10.x.x.x:8080/
2022-09-19T08:36:16.202355673Z + '[' -z http://10.x.x.x:8080/ ]
2022-09-19T08:36:16.202364107Z + var=NO_PROXY
2022-09-19T08:36:16.202378916Z + val=,localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,.svc,.cluster.local
2022-09-19T08:36:16.202388443Z + eval 'v="$NO_PROXY"'
2022-09-19T08:36:16.202412613Z + v=redacted,.redacted,.redacted,.redacted,.redacted,.redacted,10.x.x.x,10.x.x.0/24,10.x.x.0/24,10.x.x.0/24,longhorn-system,cattle-system,localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,.svc,.cluster.local
2022-09-19T08:36:16.202428391Z + '[' -z redacted,.redacted,.redacted,.redacted,.redacted,.redacted,10.x.x.x.x,10.x.x.0/24,10.x.x.0/24,10.x.x.0/24,longhorn-system,cattle-system,localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,.svc,.cluster.local ]
2022-09-19T08:36:16.202437797Z + chroot /host /var/lib/rancher/agent/tmp/tmp.7An2NIivQE/install.sh
2022-09-19T08:36:16.212819217Z [INFO]  Using default agent configuration directory /etc/rancher/agent
2022-09-19T08:36:16.212901059Z [INFO]  Using default agent var directory /var/lib/rancher/agent
2022-09-19T08:36:16.218658450Z [WARN]  /usr/local is read-only or a mount point; installing to /opt/rancher-system-agent
2022-09-19T08:36:31.600918129Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
2022-09-19T08:36:52.081757796Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
2022-09-19T08:37:12.565070754Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
2022-09-19T08:37:33.042196553Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
2022-09-19T08:37:53.528619959Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
2022-09-19T08:38:14.002510749Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
2022-09-19T08:38:34.481101439Z [ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
```

**Comment by starbops** [comment]
Hi @bashofmann, we have spotted the root cause. This is particularly triggered by the http proxy settings. The readiness probe tries to hit the endpoint with HTTP GET request, and the request is being proxied by the proxy server, which is not able to access the repo VM, so the readiness probe fails again and again.

We set up a Squid server and a Harvester v1.0.2 VM with http(s)-server pointed to the Squid server. After triggering the upgrade to v1.0.3, we can see the logs from the access.log of the Squid server:
<img width="1283" alt="Screen Shot 2022-09-21 at 11 47 57" src="https://user-images.githubusercontent.com/1827717/191413420-022970a3-1fc8-4e32-85db-467b16d080ad.png">

And the readiness fails, hence the repo VM can't start normally:
<img width="1269" alt="Screen Shot 2022-09-21 at 11 48 18" src="https://user-images.githubusercontent.com/1827717/191413538-240aeb8a-ab59-438b-b7d5-47697188dcb4.png">

To work around this in the user's current environment, `harvester-system` needs to be added into the `noProxy` setting like the following:
<img width="559" alt="Screen Shot 2022-09-21 at 12 22 06" src="https://user-images.githubusercontent.com/1827717/191414220-6fa5795b-0b1f-44db-9e2f-8cad1e199d69.png">

We will also append `harvester-system` to the user configured no-proxy to ensure the internal traffic works, just like #2436.

**Comment by TachunLin** [comment]
### Test plan 1
Setup airgapped environment with proxy server, **do not set** the `no proxy` attribute in Harvester `http-proxy` 
And perform upgrade according to https://github.com/harvester/harvester/pull/2808#issue-1381869632 
The purpose is testing whether Harvester can automatically add the no proxy settings. 

### Test plan 2 
Setup airgapped environment with proxy server, **set** the related `no proxy` attribute in Harvester `http-proxy` 
And integrate Harvester to Rancher according to the existing test plan https://harvester.github.io/tests/manual/harvester-rancher/38-import-airgapped-harvester-from-airgapped-rancher/

**Comment by TachunLin** [comment]
Discussed we will verify this issue on v1.1 branch
Move to ready for testing and verify after related PR https://github.com/harvester/harvester/pull/3296 have been merged

**Comment by TachunLin** [comment]
### Test Plan 1 Result: Failed 
1. By adding the following proxy settings in Harvester (did not include `no proxy` value) 
  ```
  {
    "httpProxy": "http://192.168.0.254:3128",
    "httpsProxy": "http://192.168.0.254:3128",
  }
  
  ```
2. Then atart upgrade from v1.1 master to v1.1 master

3. We did not encounter the vm display `unhealthy` status 

4. But we **failed** to upgrade, stuck at Upgrading node with error `Job was active longer than specific deadline
  ![image](https://user-images.githubusercontent.com/29251855/208859646-ee48fd6f-42a3-4348-8cd3-362d97a9b686.png)
  ![image](https://user-images.githubusercontent.com/29251855/208926967-f2778acc-1119-4b32-9307-92f87fdf234c.png)

5. Found the upgrade pod encounter `ImagePullBackOff` error
![image](https://user-images.githubusercontent.com/29251855/208858897-97c8007b-7ca7-4235-ba32-3d2cb7023063.png)

#### Support bundle: 
[supportbundle_7caabd91-015c-4d35-b858-c934eecd0316_2022-12-21T14-24-27Z.zip](https://github.com/harvester/harvester/files/10278466/supportbundle_7caabd91-015c-4d35-b858-c934eecd0316_2022-12-21T14-24-27Z.zip)

**Comment by TachunLin** [comment]
Trying to add `no proxy` settings back and rerun the **test plan 1.** 

  ```
  {
    "httpProxy": "http://192.168.0.254:3128",
    "httpsProxy": "http://192.168.0.254:3128",
    "noProxy": "localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,.svc,.cluster.local,example.com, cattle-system, longhorn-system"
  }
  ```

But still got the `ImagePullBackOff` error
![image](https://user-images.githubusercontent.com/29251855/208946960-774f0f73-fc68-428a-b7a5-7107540a027e.png)

![image](https://user-images.githubusercontent.com/29251855/208946919-9b7ba6b0-757c-418d-b60a-65f4ca059a12.png)

#### support bundle:
[supportbundle_e94450e9-0212-4f9b-b6ef-6ee95abd5ee4_2022-12-21T15-59-49Z.zip](https://github.com/harvester/harvester/files/10279263/supportbundle_e94450e9-0212-4f9b-b6ef-6ee95abd5ee4_2022-12-21T15-59-49Z.zip)

**Comment by TachunLin** [comment]
### Test Plan 1 Rerun Result: PASS

For dev build only, we need to update the upgrade image to the required tag name
  ```
  ctr -n k8s.io i tag docker.io/rancher/harvester-upgrade:26c395ac docker.io/rancher/harvester-upgrade:v1.1-26c395ac-head
  ```

Tested adding the following proxy setting with `noProxy` attribute 
  ```
  {
    "httpProxy": "http://192.168.0.254:3128",
    "httpsProxy": "http://192.168.0.254:3128",
    "noProxy": "localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,.svc,.cluster.local,example.com, cattle-system, longhorn-system"
  }
  ```

And also tested without having the `noProxy` attribute 
  ```
  {
    "httpProxy": "http://192.168.0.254:3128",
    "httpsProxy": "http://192.168.0.254:3128",
  }
  ```

For either ways we can successfully complete the Harvester upgrade 
  ![image](https://user-images.githubusercontent.com/29251855/209085908-1696e596-d98c-43bd-a8ff-8d42ed6d41b7.png)

**Comment by TachunLin** [comment]
### Test Plan 2 Result: PASS

In the airgapped environment where Harvester `v1.1 head` have been integrated with Rancher `v2.7.0`  

When we add the following proxy settings in Harvester
  ```
  {
    "httpProxy": "http://192.168.0.254:3128",
    "httpsProxy": "http://192.168.0.254:3128",
    "noProxy": "localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,.svc,.cluster.local,example.com, cattle-system, longhorn-system"
  }
  ```

And follow the steps to provision RKE2 guest cluster https://github.com/harvester/harvester/issues/1052#issuecomment-997620981 
  ![image](https://user-images.githubusercontent.com/29251855/209123288-26b74f64-3f40-4e7e-8776-1024788fab56.png)


We can upgrade successfully, did not enounter the upgrade-repo VM `Unhealthy` state 
After upgrade the guest cluster VM will automatically been shut downded.
  ![image](https://user-images.githubusercontent.com/29251855/209124986-48f804dc-62a3-4cde-9f82-077efe4ed45b.png)

After we manually start the VM, the RKE2 cluster can recover to `running` state
  ![image](https://user-images.githubusercontent.com/29251855/209127865-9c01d6bb-57b7-4413-b25a-47ca8842d7ae.png)

And the guest VM can work as expected 
![image](https://user-images.githubusercontent.com/29251855/209129690-545dca2a-d58d-4249-96e2-23c3cc6e4adf.png)

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1-26c395ac-head` .Close this issue.

### Result

- **Pass** the test plan 1 as result: https://github.com/harvester/harvester/issues/2782#issuecomment-1362525952
- **Pass** the test plan 2 as result: https://github.com/harvester/harvester/issues/2782#issuecomment-1362755325 

### Test Information
* Test Environment: 1 node Harvester on local kvm machine
* Harvester version: v1.1-26c395ac-head
* Rancher version: v2.7.0

### Verify Steps
Test plan 1: https://github.com/harvester/harvester/pull/2808#issue-1381869632
Test plan 2: https://github.com/harvester/harvester/issues/1052#issuecomment-997620981 

### Additional Context 
Please remember to use the following docker script to startup Rancher container instance with proxy support 
  ```
  sudo docker run -d --restart=unless-stopped -p 80:80 -p 443:443 -e HTTP_PROXY="http://192.168.0.254:3128" -e HTTPS_PROXY="http://192.168.0.254:3128" -e NO_PROXY="localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,192.168.0.0/16,.svc,.cluster.local,example.com, cattle-system, longhorn-system, harvester-system" --privileged rancher/rancher:v2.7.0
  ```
